packages/actor-extract-links-extract-tree/lib/SolutionDomain.ts

```diff
@@ -0,0 +1,198 @@
+import { SolutionRange } from './SolutionRange';
+import { LogicOperator } from './solverInterfaces';
+
+const AN_EMPTY_SOLUTION_RANGE = new SolutionRange(undefined);
+/**
+ * A class representing the domain of a solution of system of boolean equation.
+ * Every operation return a new object.
+ */
+export class SolutionDomain {
+  /**
+     * The multiple segment of the domain, it is always order by the lower bound
+     * of the SolutionRange.
+     */
+  private domain: SolutionRange[] = [];
+
+  /**
+    * The last operation apply to the domain
+    *
+    */
+  private lastOperation: LogicOperator | undefined = undefined;
+
+  /**
+     * Get the multiple segment of the domain.
+     * @returns {SolutionRange[]}
+     */
+  public get_domain(): SolutionRange[] {
+    return new Array(...this.domain);
+  }
+
+  /**
+     * Check whether the domain is empty
+     * @returns {boolean} Return true if the domain is empty
+     */
+  public isDomainEmpty(): boolean {
+    return this.domain.length === 0;
+  }
+
+  /**
+     * Create a new SolutionDomain with an inititial value.
+     * @param {SolutionRange} initialRange
+     * @returns {SolutionDomain}
+     */
+  public static newWithInitialValue(initialRange: SolutionRange): SolutionDomain {
+    const newSolutionDomain = new SolutionDomain();
+    if (!initialRange.isEmpty) {
+      newSolutionDomain.domain = [ initialRange ];
+    }
+    return newSolutionDomain;
+  }
+
+  /**
+     * Clone the current solution domain.
+     * @returns {SolutionDomain}
+     */
+  public clone(): SolutionDomain {
+    const newSolutionDomain = new SolutionDomain();
+    newSolutionDomain.domain = new Array(...this.domain);
+    newSolutionDomain.lastOperation = this.lastOperation;
+    return newSolutionDomain;
+  }
+
+  /**
+     * Modifify the current solution range by applying a logical operation.
+     * @param {SolutionRange} range - The range of the incoming operation to apply.
+     * @param {LogicOperator} operator - The logical operation to apply.
+     * @returns {SolutionDomain} - A new SolutionDomain with the operation applied.
+     */
+  public add({ range, operator }: { range?: SolutionRange; operator: LogicOperator }): SolutionDomain {
```

Comment:
Why is this implemented in the domain class?
Wouldn't it be easier to have different operators (and, or, not, ...), and implement something like an `apply` method on them, which accepts one or more SolutionDomain's, and produces a new SolutionDomain?